File: assets/codebase/rendering/DirectionalShadowCameraOrbitComponent.cs
namespace city.rendering {
    /// <summary>
    /// Keeps the showcase camera on a slow elevated orbit while always looking back toward the plaza center.
    /// </summary>
    public sealed class DirectionalShadowCameraOrbitComponent : UpdateComponent {
        /// <summary>
        /// Gets or sets the world-space orbit center.
        /// </summary>
        public float3 OrbitCenter { get; set; }

        /// <summary>
        /// Gets or sets the orbit radius in world units.
        /// </summary>
        public float OrbitRadius { get; set; }

        /// <summary>
        /// Gets or sets the vertical offset applied relative to the orbit center.
        /// </summary>
        public float OrbitHeight { get; set; }

        /// <summary>
        /// Gets or sets the base orbit angle in radians.
        /// </summary>
        public float BaseAngleRadians { get; set; }

        /// <summary>
        /// Gets or sets the angular speed in radians per second.
        /// </summary>
        public float AngularSpeedRadians { get; set; }

        /// <summary>
        /// Gets or sets the fixed downward camera pitch in radians.
        /// </summary>
        public float LookDownPitchRadians { get; set; }

        /// <summary>
        /// Evaluates the current camera orbit position and orientation from deterministic absolute runtime time.
        /// </summary>
        public override void Update() {
            base.Update();

            double angleRadians = BaseAngleRadians + (AngularSpeedRadians * Core.Instance.TotalElapsedSeconds);
            double x = OrbitCenter.X + (Math.Sin(angleRadians) * OrbitRadius);
            double z = OrbitCenter.Z + (Math.Cos(angleRadians) * OrbitRadius);
            Parent.LocalPosition = new float3((float)x, OrbitCenter.Y + OrbitHeight, (float)z);

            float4 orientation;
            float4.CreateFromYawPitchRoll((float)angleRadians, LookDownPitchRadians, 0f, out orientation);
            orientation.Normalize();
            Parent.LocalOrientation = orientation;
        }
    }
}

File: windows-build-20260602-cube-test-ground-high-direct-fix2/cooked/code/gameplay/DirectionalShadowCameraOrbitComponent.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "DirectionalShadowCameraOrbitComponent.hpp"
#include "system/math.hpp"
#include "DirectionalShadowCameraOrbitComponent.hpp"
#include "Component.hpp"
#include "Core.hpp"
#include "Entity.hpp"
#include "UpdateComponent.hpp"
#include "float3.hpp"
#include "float4.hpp"
#include "system/math.hpp"

DirectionalShadowCameraOrbitComponent::DirectionalShadowCameraOrbitComponent() : AngularSpeedRadians(), BaseAngleRadians(), LookDownPitchRadians(), OrbitCenter(), OrbitHeight(), OrbitRadius()
{
}

float DirectionalShadowCameraOrbitComponent::get_AngularSpeedRadians()
{
return this->AngularSpeedRadians;
}

void DirectionalShadowCameraOrbitComponent::set_AngularSpeedRadians(float value)
{
this->AngularSpeedRadians = value;
}

float DirectionalShadowCameraOrbitComponent::get_BaseAngleRadians()
{
return this->BaseAngleRadians;
}

void DirectionalShadowCameraOrbitComponent::set_BaseAngleRadians(float value)
{
this->BaseAngleRadians = value;
}

float DirectionalShadowCameraOrbitComponent::get_LookDownPitchRadians()
{
return this->LookDownPitchRadians;
}

void DirectionalShadowCameraOrbitComponent::set_LookDownPitchRadians(float value)
{
this->LookDownPitchRadians = value;
}

float3 DirectionalShadowCameraOrbitComponent::get_OrbitCenter()
{
return this->OrbitCenter;
}

void DirectionalShadowCameraOrbitComponent::set_OrbitCenter(float3 value)
{
this->OrbitCenter = value;
}

float DirectionalShadowCameraOrbitComponent::get_OrbitHeight()
{
return this->OrbitHeight;
}

void DirectionalShadowCameraOrbitComponent::set_OrbitHeight(float value)
{
this->OrbitHeight = value;
}

float DirectionalShadowCameraOrbitComponent::get_OrbitRadius()
{
return this->OrbitRadius;
}

void DirectionalShadowCameraOrbitComponent::set_OrbitRadius(float value)
{
this->OrbitRadius = value;
}

void DirectionalShadowCameraOrbitComponent::Update()
{
UpdateComponent::Update();
const double angleRadians = this->BaseAngleRadians + (this->AngularSpeedRadians * Core::get_Instance()->get_TotalElapsedSeconds());
const double x = this->OrbitCenter.X + (Math::Sin(angleRadians) * this->OrbitRadius);
const double z = this->OrbitCenter.Z + (Math::Cos(angleRadians) * this->OrbitRadius);
this->Parent->set_LocalPosition(float3(static_cast<float>(x), this->OrbitCenter.Y + this->OrbitHeight, static_cast<float>(z)));
float4 orientation;
float4::CreateFromYawPitchRoll__out3(static_cast<float>(angleRadians), this->LookDownPitchRadians, 0.0f, orientation);
orientation.Normalize();
this->Parent->set_LocalOrientation(orientation);
}

